x/x.go

```diff
@@ -1019,6 +1021,85 @@ func IsGuardian(groups []string) bool {
 	return false
 }
 
+// MonitorCacheHealth periodically monitors the cache metrics and reports if
+// there is high contention in the cache.
+func MonitorCacheHealth(period time.Duration, prefix string, db *badger.DB, closer *z.Closer) {
+	defer closer.Done()
+
+	getMetrics := func(ct string) *ristretto.Metrics {
+		var metrics *ristretto.Metrics
+		switch ct {
+		case "pstore-block":
+			metrics = db.BlockCacheMetrics()
+		case "pstore-index":
+			metrics = db.IndexCacheMetrics()
+		case "WALstore-block":
+			metrics = db.BlockCacheMetrics()
+		case "WALstore-index":
+			metrics = db.IndexCacheMetrics()
+		default:
+			metrics = nil
+		}
+		return metrics
+	}
+
+	checkCache := func(ct string) {
+		metrics := getMetrics(ct)
+		if metrics == nil {
+			return
+		}
+		switch ct {
+		case "pstore-block":
+			ostats.Record(context.Background(), PBlockHitRatio.M(metrics.Ratio()))
+		case "pstore-index":
+			ostats.Record(context.Background(), PIndexHitRatio.M(metrics.Ratio()))
+		case "WALstore-block":
+			ostats.Record(context.Background(), WBlockHitRatio.M(metrics.Ratio()))
+		case "WALstore-index":
+			ostats.Record(context.Background(), WIndexHitRatio.M(metrics.Ratio()))
+		default:
+			panic("invalid cache type")
+		}
+
+		// If the mean life expectancy is less than 10 seconds, the cache
+		// might be too small.
+		le := metrics.LifeExpectancySeconds()
+		lifeTooShort := le.Count > 0 && float64(le.Sum)/float64(le.Count) < 10
+		hitRatioTooLow := metrics.Ratio() < 0.4
+		if bool(glog.V(2)) && (lifeTooShort || hitRatioTooLow) {
+			glog.Warningf("======== Cache might be too small %s =====", ct)
+			glog.Warningf("Metric: %+v", metrics)
+			glog.Warningf("Life expectancy: %+v", le)
+		}
+	}
+
+	logMetrics := func(ct string) {
+		if metrics := getMetrics(ct); metrics != nil {
+			prefix := fmt.Sprintf("%s-%s", prefix, ct)
+			le := metrics.LifeExpectancySeconds()
+			glog.V(2).Infof("%s metrics %+v %+v", prefix, metrics, le)
+		}
+	}
+
+	ticker := time.NewTicker(period)
+	tickerLog := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+	defer ticker.Stop()
```

Comment:
Should be tickerLog. Try to keep the defer statement near the initialization.

Reply:
Fixed. Thanks.